### ISSUE TITLE
query/store/receive/rule: Consider splitting StoreResponse frames to max 1MB.

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#2332]
We do that only on sidecar (well, remote read does that for us).

It would be nice to benchmark, see if this matters, but gRPC recommends max 1MB frames. We might get outside of that during heavy queries.

**Comment by yeya24** [comment]
I can take this one. But let me make this more clear. What we need to do here is to implement something like `encodeChunks` to make the gRPC response max to 1MB. Then do some benchmarks to it.
But do I need to apply this to all `Store` methods other than `Series`? I guess the response of `LabelValues` cannot exceed 1MB

**Comment by bwplotka** [comment]
That's correct. :+1: 

It would be awesome if you can create/run some microbenchmarks to confirm the difference (: This might involve creating new one, as we currently have only one for [store](https://github.com/thanos-io/thanos/blob/27a4a235fa62ead7743bb443c730b511e82255ad/pkg/store/bucket_test.go#L1094) and attempts for [query](https://github.com/thanos-io/thanos/pull/2305). Without measurement, it's just blind chance, so it's important. I recommend using recently improved `[funcbench](https://github.com/prometheus/test-infra/tree/master/funcbench)` and joining our `#benchmark` slack channel (: 

LabelValues can exceed that but less likely. Definitely let's start with just `Series`

**Comment by yeya24** [comment]
I did some experiments on Store's series methods [here](https://github.com/yeya24/thanos/pull/26), seems there is not much difference. Or maybe the current test cases are not enough?

```
name                                                     old time/op    new time/op    delta
Series/10e6SeriesWithOneSample/1of10000000-40               4.83s ± 2%     4.90s ± 2%    ~     (p=0.222 n=5+5)
Series/10e6SeriesWithOneSample/10of10000000-40              4.86s ± 4%     4.92s ± 6%    ~     (p=0.421 n=5+5)
Series/10e6SeriesWithOneSample/100of10000000-40             4.95s ± 5%     5.07s ± 9%    ~     (p=0.151 n=5+5)
Series/10e6SeriesWithOneSample/1000of10000000-40            5.01s ± 5%     4.91s ± 1%    ~     (p=0.690 n=5+5)
Series/10e6SeriesWithOneSample/10000of10000000-40           4.99s ± 6%     4.98s ± 1%    ~     (p=0.690 n=5+5)
Series/10e6SeriesWithOneSample/100000of10000000-40          4.88s ± 2%     5.08s ± 4%  +4.27%  (p=0.008 n=5+5)
Series/10e6SeriesWithOneSample/1000000of10000000-40         6.54s ± 8%     6.40s ± 3%    ~     (p=0.548 n=5+5)
Series/OneSeriesWith100e6Samples/1of100000000-40           36.5ms ± 2%    37.5ms ± 1%  +2.74%  (p=0.008 n=5+5)
Series/OneSeriesWith100e6Samples/10of100000000-40          36.8ms ± 7%    37.5ms ± 1%    ~     (p=0.690 n=5+5)
Series/OneSeriesWith100e6Samples/100of100000000-40         36.2ms ± 5%    35.9ms ± 3%    ~     (p=0.841 n=5+5)
Series/OneSeriesWith100e6Samples/1000of100000000-40        36.3ms ± 4%    36.5ms ± 1%    ~     (p=0.841 n=5+5)
Series/OneSeriesWith100e6Samples/10000of100000000-40       37.6ms ± 3%    37.7ms ± 3%    ~     (p=1.000 n=5+5)
Series/OneSeriesWith100e6Samples/100000of100000000-40      37.6ms ± 3%    38.0ms ± 2%    ~     (p=0.690 n=5+5)
Series/OneSeriesWith100e6Samples/1000000of100000000-40     40.2ms ± 3%    40.1ms ± 2%    ~     (p=0.690 n=5+5)
Series/OneSeriesWith100e6Samples/10000000of100000000-40     129ms ± 4%     128ms ± 4%    ~     (p=0.690 n=5+5)

name                                                     old alloc/op   new alloc/op   delta
Series/10e6SeriesWithOneSample/1of10000000-40              1.58GB ± 0%    1.58GB ± 0%    ~     (p=0.690 n=5+5)
Series/10e6SeriesWithOneSample/10of10000000-40             1.58GB ± 0%    1.58GB ± 0%    ~     (p=1.000 n=5+5)
Series/10e6SeriesWithOneSample/100of10000000-40            1.58GB ± 0%    1.58GB ± 0%    ~     (p=0.548 n=5+5)
Series/10e6SeriesWithOneSample/1000of10000000-40           1.58GB ± 0%    1.58GB ± 0%    ~     (p=0.421 n=5+5)
Series/10e6SeriesWithOneSample/10000of10000000-40          1.59GB ± 0%    1.59GB ± 0%  +0.01%  (p=0.008 n=5+5)
Series/10e6SeriesWithOneSample/100000of10000000-40         1.67GB ± 0%    1.67GB ± 0%  -0.12%  (p=0.029 n=4+4)
Series/10e6SeriesWithOneSample/1000000of10000000-40        2.54GB ± 0%    2.53GB ± 1%    ~     (p=0.310 n=5+5)
Series/OneSeriesWith100e6Samples/1of100000000-40           59.3MB ± 0%    59.3MB ± 0%    ~     (p=0.730 n=4+5)
Series/OneSeriesWith100e6Samples/10of100000000-40          59.3MB ± 0%    59.3MB ± 0%    ~     (p=0.452 n=5+5)
Series/OneSeriesWith100e6Samples/100of100000000-40         59.3MB ± 0%    59.3MB ± 0%    ~     (p=0.889 n=5+5)
Series/OneSeriesWith100e6Samples/1000of100000000-40        59.2MB ± 0%    59.2MB ± 0%    ~     (p=0.841 n=5+5)
Series/OneSeriesWith100e6Samples/10000of100000000-40       59.4MB ± 0%    59.4MB ± 0%    ~     (p=0.151 n=5+5)
Series/OneSeriesWith100e6Samples/100000of100000000-40      60.4MB ± 0%    60.4MB ± 0%    ~     (p=0.690 n=5+5)
Series/OneSeriesWith100e6Samples/1000000of100000000-40     69.1MB ± 0%    69.1MB ± 0%    ~     (p=0.310 n=5+5)
Series/OneSeriesWith100e6Samples/10000000of100000000-40     298MB ± 0%     298MB ± 0%  +0.00%  (p=0.008 n=5+5)

name                                                     old allocs/op  new allocs/op  delta
Series/10e6SeriesWithOneSample/1of10000000-40               20.1M ± 0%     20.1M ± 0%    ~     (p=0.548 n=5+5)
Series/10e6SeriesWithOneSample/10of10000000-40              20.1M ± 0%     20.1M ± 0%    ~     (p=0.841 n=5+5)
Series/10e6SeriesWithOneSample/100of10000000-40             20.1M ± 0%     20.1M ± 0%    ~     (p=0.690 n=5+5)
Series/10e6SeriesWithOneSample/1000of10000000-40            20.1M ± 0%     20.1M ± 0%    ~     (p=0.548 n=5+5)
Series/10e6SeriesWithOneSample/10000of10000000-40           20.1M ± 0%     20.1M ± 0%    ~     (p=0.841 n=5+5)
Series/10e6SeriesWithOneSample/100000of10000000-40          20.6M ± 0%     20.6M ± 0%    ~     (p=1.000 n=5+5)
Series/10e6SeriesWithOneSample/1000000of10000000-40         25.1M ± 0%     25.1M ± 0%    ~     (p=1.000 n=5+5)
Series/OneSeriesWith100e6Samples/1of100000000-40              270 ± 0%       269 ± 0%  -0.37%  (p=0.029 n=4+4)
Series/OneSeriesWith100e6Samples/10of100000000-40             270 ± 0%       269 ± 0%    ~     (p=0.095 n=4+5)
Series/OneSeriesWith100e6Samples/100of100000000-40            270 ± 0%       269 ± 0%    ~     (p=0.095 n=5+4)
Series/OneSeriesWith100e6Samples/1000of100000000-40           288 ± 0%       287 ± 0%    ~     (p=0.095 n=4+5)
Series/OneSeriesWith100e6Samples/10000of100000000-40          447 ± 0%       447 ± 0%    ~     (all equal)
Series/OneSeriesWith100e6Samples/100000of100000000-40       1.98k ± 0%     1.98k ± 0%    ~     (p=0.762 n=5+5)
Series/OneSeriesWith100e6Samples/1000000of100000000-40      17.2k ± 0%     17.2k ± 0%  +0.09%  (p=0.008 n=5+5)
Series/OneSeriesWith100e6Samples/10000000of100000000-40      169k ± 0%      170k ± 0%  +0.09%  (p=0.008 n=5+5)
```

**Comment by bwplotka** [comment]
Let's talk on PR.


TL;DR the benchmark you run does not test any frame splitting results.

**Comment by bwplotka** [comment]
Looks like not worth it for now (: Thanks @yeya24 for initial benchmarks!
